### PR TITLE
feat(api): Allow treating errors as false-positives (ignore them and continue with the run)

### PR DESCRIPTION
**Pull request by SyntaxColoring**
## Overview

This does the interesting part of EXEC-676. See there for background.

## Changelog

When a command encounters a defined error, it now enacts *two* state updates:

1. The first one is what we had before. It represents all the changes that the command made—up to, and including, the point of failure. This is applied as soon as the command fails, as part of the `FailCommandAction`.
2. The second one is new. It represents *additional* changes to make *if* this error turns out to be a false positive. This is *not* applied as part of the `FailCommandAction`. It's merely stored, and applied later, when (or if) a client tells us that the error was a false positive. I think of it as a deferred state update.

For example, if a `pickUpTip` fails with a`tipPhysicallyMissing` error:

1. A tip is logically consumed from the tip rack, but no tip is logically placed on the pipette. (These are the documented semantics of the `tipPhysicallyMissing` error.) This happens via `state_update`.
2. But if a client selects "Ignore error and skip to next step," a tip is then logically placed on the pipette. You can then continue on to the rest of the protocol, and it should just work.

We unfortunately need to go out of our way to keep the hardware API's state in sync with Protocol Engine's state as we do this.

(All of the above is EXEC-785.)

We can use this to allow continuing from`tipPhysicallyMissing` and `tipPhysicallyAttached` errors in `pickUpTip`, `dropTip`, and `dropTipInPlace` commands. (EXEC-778, EXEC-779.)

## Test plan

Most easily tested with #16601.

* [x] Play around with inducing and recovering from errors. In particular, try continuing from failed tip pickups and tip drops.
* [x] Pay attention to path planning. For example, if you ignore and continue from a failed tip pickup, the pipette should move as if it *does* have a tip attached.
* [x] Make sure the "ignore all errors of this type" button works as advertised.
* [x] Make sure the older error recovery options ("retry", "cancel") still work.

## Review requests

* In general, any thoughts on the fundamental approach?
* In general, any thoughts on naming?
* See comments below for specific stuff.

## Risk assessment

Medium or maybe high. The blast radius is confined to error recovery mode, but this does do weird stuff in bad ways.

Implementing this as Protocol Engine actions instead of Protocol Engine commands bypasses some of our usual safeties. We discussed this a little in https://github.com/Opentrons/opentrons/pull/16564#discussion_r1811191155.

### CODE REVIEW

**Review by sfoster1 — COMMENTED** [review]
I think the hardware synchronizer is the only real way to do it and therefore is fine. In the long term I think we should pull more state out of the hardware controller; it's stuck around in the engine era for the sake of supporting older APIs and direct hardware controller usage, but it is probably time to split that stuff out into a compat layer.

I think a nice improvement would probably be to centralize not only the synchronizing code but also where it gets called, by having a `HardwareState` or something that eats the resume from recovery action instead of it being a side effect of the top level engine call. Those are always awful to deal with.

File: api/src/opentrons/protocol_engine/commands/drop_tip.py [hunk 1]
```diff
@@ -146,7 +146,15 @@ async def execute(self, params: DropTipParams) -> _ExecuteReturn:
                     )
                 ],
             )
-            return DefinedErrorData(public=error, state_update=state_update)
+            state_update_if_false_positive = update_types.StateUpdate()
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/commands/drop_tip.py:149]: I dont know why the false positive is so confusing to me but can we change it? `state_update_if_command_failed` if its only me we can leave as it

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/commands/drop_tip.py:149]: Discussed a bit in-person: It sounds like we're keeping the "false positive" terminology for now, but I'm definitely open to better ideas. We can change the terminology relatively easily before end of day Monday.

File: api/src/opentrons/protocol_engine/create_protocol_engine.py
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/create_protocol_engine.py:1]: This is kind of an unrelated refactor.

Protocol Engine has a bunch of injected dependencies that, themselves, need to be wired up to each other. This is usually straightforward, but it is, you know, code, and we can make mistakes. Especially when the dependencies' constructors do defaulting that can permit us to forget to wire something up.

The wire-up was split between this `create_protocol_engine()` helper and `ProtocolEngine.__init__()`, which was making me a little nervous. This moves most of it to `create_protocol_engine()`.

File: api/src/opentrons/protocol_engine/error_recovery_policy.py [hunk 2]
```diff
@@ -26,10 +26,20 @@ class ErrorRecoveryType(enum.Enum):
...
+    """Enter interactive error recovery mode."""
 
-    IGNORE_AND_CONTINUE = enum.auto()
-    """Continue with the run, as if the command never failed."""
+    CONTINUE_WITH_ERROR = enum.auto()
+    """Continue without interruption, carrying on from whatever error state the failed
+    command left the engine in.
+
+    This is like `ProtocolEngine.resume_from_recovery(reconcile_false_positive=False)`.
+    """
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/error_recovery_policy.py:36]: Any better names than `CONTINUE_WITH_ERROR`?

Would `ASSUME_TRUE_POSITIVE_AND_CONTINUE` be too silly?

File: api/src/opentrons/protocol_engine/execution/tip_handler.py [hunk 3]
```diff
@@ -414,13 +436,20 @@ async def drop_tip(
...
         """
         assert False, "TipHandler.cache_tip should not be used with virtual pipettes"
 
+    def remove_tip(self, pipette_id: str) -> None:
+        """See documentation on abstract base class.
+
+        This should not be called when using virtual pipettes.
+        """
+        assert False, "TipHandler.remove_tip should not be used with virtual pipettes"
+
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/execution/tip_handler.py:452]: I'm not fully sure if this is `assert` is the right thing to do, but this is what we do for `TipHandler.cache_tip()`, so 🤷

File: api/src/opentrons/protocol_engine/create_protocol_engine.py
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/create_protocol_engine.py:1]: This is kind of an unrelated refactor.

Protocol Engine has a bunch of injected dependencies that, themselves, need to be wired up to each other. This is usually straightforward, but it is, you know, code, and we can make mistakes. Especially when the dependencies' constructors do defaulting that can permit us to forget to wire something up.

The wire-up was split between this `create_protocol_engine()` helper and `ProtocolEngine.__init__()`, which was making me a little nervous. This moves most of it to `create_protocol_engine()`.

File: api/src/opentrons/protocol_engine/error_recovery_policy.py (hunk 2 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_engine/error_recovery_policy.py:36]: i would say `CONTINUE_FROM_RECOVERY` maybe? I think `CONTINUE_WITH_ERROR` is ~fine though.

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

File: api/src/opentrons/protocol_engine/error_recovery_policy.py (hunk 2 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_engine/error_recovery_policy.py:36]: `CONTINUE_WITH_ERROR` makes more sense to me

File: api/src/opentrons/protocol_engine/execution/error_recovery_hardware_state_synchronizer.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    tip geometry every time we call a hardware API movement method, instead of
+    just once when we pick up a tip. Use Protocol Engine state as the single source
+    of truth.
+    """
+
+    def __init__(self, hardware_api: HardwareControlAPI, state_view: StateView) -> None:
+        self._hardware_api = hardware_api
+        self._state_view = state_view
+
+    def handle_action(self, action: Action) -> None:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/execution/error_recovery_hardware_state_synchronizer.py:61]: LOVELY!

**Review by TamarZanzouri — APPROVED** [review]
Very nice work! I love the refactor for the hardware synchronizer!

**Pull request merged**